Implementation of Window Functions

sdanyliv opened the pull request:
Almost done Window Functions support #542.
**What have to be done:**
- [x] Tests on different databases (fully Oracle, partially SQL Server, PostgreSQL)
- [x] Tests for all functions
- [x] Change the way how to define such functions. Some have exotic syntax
- [x] Ready to merge

As specification i have choose ORACLE, so probably many functions will not work on other databases.

jogibear9988 commented:
will they also work on old sql server, where rownumber is already used for take & skip, when my query contains take or skip?

sdanyliv commented:
@jogibear9988, not yet. But it first big step to solve such kind of issues.

ili commented:
We should do our best to write more tests before merging

ili commented on Source/Sql/Sql.ExpressionAttribute.cs:
```diff
@@ -55,6 +54,7 @@ public ExpressionAttribute(string configuration, string expression, params int[]
 			public bool   InlineParameters { get; set; }
 			public bool   ExpectExpression { get; set; }
 			public bool   IsPredicate      { get; set; }
+			public bool   HasPureConvertor { get; set; }
```
I think it is better to remove this property, you see you can set it to true but you can not implement `GetExpression`. For ex if somebody will do: `[Sql.Expression(HasPureConvertor = true` `NIE` would be thrown in runtime. So I suggest to remove this parameter and use `GetExpression(...) != null` instead of it

sdanyliv replied:
Agree

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,438 @@
+
+		public interface ISqExtensionBuilder
+		{
+			string		   Configuration    { get; }
```
formatting

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,438 @@
+					[CanBeNull] MethodInfo                       method,
+					[NotNull]   Expression[]                     arguments)
+				{
+					if (extension == null) throw new ArgumentNullException("extension");
```
formatting

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,438 @@
+
+				public SqlExtension.ExtensionParam AddParameter(string name, ISqlExpression expr)
+				{
+					return Extension.AddParameter(name, expr);;
```
NOP ;;

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,438 @@
+				ServerSideOnly   = true;
+				PreferServerSide = true;
+				ExpectExpression = true;
+				HasPureConvertor = true;
```
can remove second sets

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,438 @@
+
+			protected class ConvertHelper
+			{
+				readonly Func<Expression, ISqlExpression> _converter;
```
formatting

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,438 @@
+			public override ISqlExpression GetExpression(Expression expression, Func<Expression, ISqlExpression> converter)
+			{
+				var helper = new ConvertHelper(converter);
+				var chain = BuildFunctionsChain(expression, helper);
```
formatting

ili commented on Source/SqlProvider/BasicSqlBuilder.cs:
```diff
@@ -2119,6 +2124,69 @@ void BuildFunction(string name, ISqlExpression[] exprs)
+		
+		protected virtual void BuildSqlExtension(SqlExtension extension)
+		{
+			var sb = new StringBuilder();
```
table formatting

ili commented on Source/SqlQuery/SqlExtension.cs:
```diff
@@ -0,0 +1,297 @@
+		{
+			public ExtensionParam(string name, ISqlExpression expression)
+			{
+				Name = name;
```
table formatting

ili commented on Source/SqlQuery/SqlExtension.cs:
```diff
@@ -0,0 +1,297 @@
+			public QueryElementType ElementType { get { return QueryElementType.SqlExtensionParam; } }
+			public StringBuilder ToString(StringBuilder sb, Dictionary<IQueryElement, IQueryElement> dic)
+			{
+				//TODO: Impelement
```
really needed or can be removed?

sdanyliv replied:
It will be good for debugging in the future.

ili commented on Source/SqlQuery/SqlExtension.cs:
```diff
@@ -0,0 +1,297 @@
+		public Type             SystemType { get;         set; }
+		public string           Expr       { get;         set; }
+		public int              Precedence { get; private set; }
+		public bool             IsRoot     { get; set;         }
```
formatting

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,638 @@
+			}
+
+			public string BuilderMethod   { get; set; }
+			public Type   BuilderType     { get; set; }
```
lets avoid using BuilderMethod & BuilderType, lets introduce 
```cs
interface IExtensionCallBuilder
{
    Build(Sql.ISqExtensionBuilder builder)
}
```
and provide type as call builder where needed

ili commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,638 @@
+
+				var extNames = Names == null ? new string[] {""} : Names.Split(',');
+
+				return extNames.Select(n => result.Expression != null ? new SqlExtensionParam(n, result.Expression) : new SqlExtensionParam(n, result.Extension));
```
looks like result is always the same regardless of `result.Expression != null`

sdanyliv commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -0,0 +1,638 @@
+				}
+
+				chains.Reverse();
+				return chains;
```
No, order i matter, especially for order by sequence.

MaceWindu commented:
Guys, don't merge it yet please. I want to test one thing related to this functionality today (https://github.com/linq2db/linq2db/issues/133)

sdanyliv commented:
It should work, rewrite your query in this way
```cs
var query = db.GetTable<AggregatesDbModel>().GroupBy(x => x.Int)
                .Select(
                x => new
                {
                    Item = x.Key.Value,
                    Value = x.Count() * 100 / Sql.Ext.Sum(x.Count()).Over().ToValue(),
                    Avg = x.Average(_ => _.Int)

                }).Where(_ => _.Avg > 5);
```

MaceWindu commented:
Thanks for example. Unfortunatelly this issue still exists.
Could you add following tests to IssueTests?

```
[Sql.Expression("COUNT(*) * 100E0 / SUM(COUNT(*)) OVER()", ServerSideOnly = true)]
		private static double CountPercents()
		{
			throw new InvalidOperationException("This function should be used only in database code");
		}

		[Test, DataContextSource]
		public void Issue133_Negative_Where(string context)
		{
			using (var db = GetDataContext(context))
			{
				var result = db.Child
					.GroupBy(_ => _.ParentID)
					.Select(_ => new { CountPercents = CountPercents(), Sum = _.Sum(r => r.ParentID) })
					.Where(_ => _.Sum != 36)
					.ToList();

				Assert.AreEqual(5, result.Count);

				// must be 100
				Assert.AreEqual(64.7058d, result.Sum(_ => _.CountPercents), 0.001);
			}
		}

		[Test, DataContextSource]
		public void Issue133_Positive_Having(string context)
		{
			using (var db = GetDataContext(context))
			{
				var result = db.Child
					.GroupBy(_ => _.ParentID)
					.Select(_ => new { CountPercents = CountPercents(), Sum = _.Sum(r => r.ParentID) })
					.Having(_ => _.Sum != 36)
					.ToList();

				Assert.AreEqual(5, result.Count);
				Assert.AreEqual(100d, result.Sum(_ => _.CountPercents), 0.001);
			}
		}

		[Test, DataContextSource]
		public void Issue133_Positive_WindowFunctions_Where(string context)
		{
			using (var db = GetDataContext(context))
			{
				var result = db.Child
					.GroupBy(_ => _.ParentID)
					.Select(_ => new
					{
						CountPercents = _.Count() * 100d / Sql.Ext.Sum(_.Count()).Over().ToValue(),
						Sum = _.Sum(r => r.ParentID)
					})
					.Where(_ => _.Sum != 36)
					.ToList();

				Assert.AreEqual(5, result.Count);
				Assert.AreEqual(100d, result.Sum(_ => _.CountPercents), 0.001);
			}
		}

		[Test, DataContextSource]
		public void Issue133_Positive_WindowFunctions_Having(string context)
		{
			using (var db = GetDataContext(context))
			{
				var result = db.Child
					.GroupBy(_ => _.ParentID)
					.Select(_ => new
					{
						CountPercents = _.Count() * 100d / Sql.Ext.Sum(_.Count()).Over().ToValue(),
						Sum = _.Sum(r => r.ParentID)
					})
					.Having(_ => _.Sum != 36)
					.ToList();

				Assert.AreEqual(5, result.Count);
				Assert.AreEqual(100d, result.Sum(_ => _.CountPercents), 0.001);
			}
		}
```
Issue133_Positive_WindowFunctions_Where behaves the same way as  Issue133_Negative_Where...

sdanyliv commented:
@MaceWindu, i'll handle this

sdanyliv commented:
@MaceWindu, looks like it is is not part of this issue. BUT, please check this small change in **WhereBuilder.cs**
```cs
		protected override IBuildContext BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
		{
			var sequence  = builder.BuildSequence(new BuildInfo(buildInfo, methodCall.Arguments[0]));
			var isHaving  = methodCall.Method.Name == "Having" || !sequence.SelectQuery.Select.GroupBy.IsEmpty;
			var condition = (LambdaExpression)methodCall.Arguments[1].Unwrap();
```
I've changed **isHaving** calculation.
What i need in that case - linq query that may fail after this "fix", because our tests works fine with this change. Please help in investigation. Thanks.

MaceWindu commented:
You are right - this is not directly related to this issue. I just thought that with native support for window functions we will be able to perform required analysis, mentioned in this comment https://github.com/linq2db/linq2db/issues/133#issuecomment-66665950 by Igor. But I don't know right now what kind of check it must be.

I think answer to your question could be found in ExpressionBuilder.CheckSubQueryForWhere method. I will analyze it later this week.
For now we can ignore this issue and fix it later as a separate PR as it is not a regression.

sdanyliv commented:
I've checked CheckSubQueryForWhere - it does not have information about Grouping. So it's why i've added this line. Looks like it works but maybe i'm missing something.

I've added your unit tests in new PR #642

MaceWindu commented on Tests/Linq/Linq/SqlExtensionTests.cs:
```diff
@@ -76,26 +76,32 @@ class DatePartBuilderPostgre: Sql.IExtensionCallBuilder
 
-				builder.AddEpression("part", partStr);
+				if (partStr != null)
+					builder.AddEpression("part", partStr);
```
"Epression" typo?

sdanyliv replied:
O yeah, tnx

MaceWindu commented on Source/Sql/Sql.ExtensionAttribute.cs:
```diff
@@ -277,18 +278,12 @@ public SqlExtensionParam AddParameter(string name, ISqlExpression expr)
-				TokenName        = tokenName;
-
 				ServerSideOnly   = true;
 				PreferServerSide = true;
```
```
ServerSideOnly   = true;
PreferServerSide = true;
```
specified twice
